File: Leetcode/src/com/leetcode/stackPriorityQueue/_227_BasicCalculatorII.java
package com.leetcode.stackPriorityQueue;

import java.util.Stack;

public class _227_BasicCalculatorII {
    public static int calculate(String s) {
        if (s == null || s.length() == 0) return 0;
        Stack<Integer> stack = new Stack<>();
        int res = 0;
        char sign = '+';// 一开始就有一个'+'
        int num = 0;
        for (int i = 0; i < s.length(); i++) {
            if (Character.isDigit(s.charAt(i))) {
                num = s.charAt(i) - '0';
                while (i + 1 < s.length() && Character.isDigit(s.charAt(i + 1))){
                    num = num * 10 + s.charAt(i + 1) - '0';
                    i++;
                }
            }
            // 这里也是if是因为需要i= i+1的情况
            // 相当于先都把这些加入 然后再判断
            if (!Character.isDigit(s.charAt(i)) && s.charAt(i) != ' '
                    || i == s.length() - 1) {
                if (sign == '+') stack.push(num);
                if (sign == '-') stack.push(-num);
                if (sign == '*') stack.push(stack.pop() * num);
                if (sign == '/') stack.push(stack.pop() / num);
                sign = s.charAt(i);
                num = 0;
            }
        }
        for (int i : stack) {
            res += i;
        }
        return res;
    }

    public static void main(String[] args) {
        calculate("3+2*2");
    }
}

File: Leetcode/src/com/leetcode/stack_priority_queue/Calculator/_227_basic_calculator_ii.py
class Solution:
    def calculate(self, s: str) -> int:
        n = len(s)
        curr_res = 0
        res = 0
        num = 0
        sign = "+"
        for i in range(n):
            ch = s[i]
            if ch.isdigit():
                num = num * 10 + int(ch)

            if i == n - 1 or ch in ("+", "-", "*", "/"):
                if sign == "+":
                    curr_res += num
                elif sign == "-":
                    curr_res -= num
                elif sign == "*":
                    curr_res *= num
                elif sign == "/":
                    # -3//2 -> -2
                    curr_res = int(curr_res / num)

                if i == n - 1 or ch in ("+", "-"):
                    res += curr_res
                    curr_res = 0

                sign = ch
                num = 0

        return res
